[ACP-192] Streaming Asynchronous Execution (#194)
avalanche-foundation/ACPs: ACPs/194-streaming-asynchronous-execution/README.md

A concurrent execution stream is responsible for clearing the queue and reporting a delayed state root for recording by later rounds of consensus.
Validation of transactions to be pushed to the queue is lightweight but guarantees eventual execution.

## Motivation

### Performance improvements

1. Concurrent consensus and execution streams eliminate node context switching, reducing latency caused by each waiting on the other.
In particular, "VM time" (akin to CPU time) more closely aligns with wall time since it is no longer eroded by consensus.
This increases gas per wall-second even without an increase in gas per VM-second.
2.  Lean, execution-only clients can rapidly execute the queue agreed upon by consensus, providing accelerated receipt issuance and state computation.
Without the need to compute state _roots_, such clients can eschew expensive Merkle data structures.
End users see expedited but identical transaction results.
3. Irregular stop-the-world events like database compaction are amortised over multiple blocks.

> 4. EOA-to-EOA transfers of value can be considered settled immediately upon being pushed to the queue by consensus.

Author: Resolved in ARR4N/ACPs#1

> 5. Introduces bursty throughput under temporary high load, preemptively accepting transactions that may take marginally longer to execute, without a reduction in security guarantees.

Author: Resolved with [this conversation](https://github.com/ARR4N/ACPs/pull/1#discussion_r2087455861).

> 5. Introduces bursty throughput under temporary high load, preemptively accepting transactions that may take marginally longer to execute, without a reduction in security guarantees.

Author: Resolved in ARR4N/ACPs#1

### User stories

1. A sophisticated DeFi trader runs a highly optimised execution client, locally clearing the transaction queue well in advance of the network—setting the stage for HFT DeFi.
2. A custodial platform filters the queue for only those transactions sent to one of their EOAs, immediately crediting user balances.

## Specification

### Transaction queue

#### Pushing transactions

Standard, synchronous execution couples a block to the ordered set of transactions to be executed in said block.
Instead, let there be a FIFO queue of all known, valid transactions—the specifics of "validity" will be addressed later.
Under SAE, a block's ordered transactions instead constitute those to be pushed to the queue

> A transaction is said to be _included_ when a block builder adds it to a block and _accepted_ once consensus, performed over blocks, agrees to add it to the queue.

Reviewer: > A transaction is said to be _included_ when a block builder adds it to a block and _accepted_ once consensus, performed over blocks, agrees to add it to the queue.

This has the added benefit that we can incorporate verifiable randomness seamlessly, since a transaction is only added to the queue once consensus totally orders it, and by then avalanchego will have its randomness. 

In the current way Avalanchego works, the transactions are executed before consensus, so the randomness is not yet available.

Author: What's the VRF in such a scenario?

> We refer to the ordered set of enqueued transactions as a _tranche_.

Author: Nope, it's then unnecessary.

#### Named blocks

The Ethereum Mainnet APIs allow for retrieving blocks by named parameters that the API server resolves based on their consensus mechanism.
Other than the _earliest_ (genesis) named block, which MUST be interpreted in the same manner, all other named blocks are mapped to SAE in terms of the execution status of all block transactions and MUST be interpreted as follows:

 * _pending_: the most recently _accepted_ block;
 * _latest_: the block that was most recently _executed_;
 * _safe_ and _finalized_: the block that was most recently _settled_.

> [!NOTE]
> The finality guarantees of Snowman consensus remove any distinction between _safe_ and _finalized_. 
> Furthermore, the _latest_ block is not at risk of re-org, only of a negligible risk of data corruption local to the API node.

### Optimal block building

As EOA-to-EOA transfers of value are entirely guaranteed upon _acceptance_, block builders SHOULD prioritise other transactions for earlier execution.
A reliable marker of such transactions is a gas limit of 21,000 as this is an indication from the sender that they do not intend to execute bytecode.

> Checking for code at the destination address is both unnecessary and unreliable as the address might be that of a contract created earlier in the queue.

Author: Resolved in ARR4N/ACPs#1

## Backwards Compatibility

All ACPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The ACP should provide recommendations for dealing with these incompatibilities.

## Reference Implementation

> A reference/example implementation that people can use to assist in understanding or implementing this specification. If the implementation is too large to reasonably be included inline, then consider adding it to the ACP subdirectory or linking to a PR in an external repository.

Author: Resolved in ARR4N/ACPs#1

> ## Security Considerations

Reviewer 2: I think everything currently in the Security Considerations section needs to be moved into the Specification section.

The Security Considerations section should talk about the tradeoffs made (including the calculations of the worst case DoS multiplier and argument that the multiplier on the fee isn't that bad). The Specification section should describe what verification needs to take place (so that the DoS multiplier is bounded).

### Worst-case transaction validity

To avoid a DoS vulnerability on execution, we require an upper bound on transaction gas cost (i.e. amount $\times$ price).
Beyond regular requirements for transaction validity (e.g. nonce, signature, etc.) we therefore introduce "worst-case cost" validity.

Consider a hypothetical situation in which all enqueued transactions use their entire gas limit during execution.
We can prove that, for every transaction, this would result in the greatest possible:

1. Consumption of gas units (by definition of the gas limit); and
2. Gas excess $x$ (and therefore gas price) at the time of execution.

> For a queue of transactions $Q = \\{i\\}_ {i \ge 0}$ the gas excess $x_j$ immediately prior to execution of transaction $j \in Q$ is a monotonic, non-decreasing function of the gas usage of all preceding transactions in the queue; i.e. $x_j~:=~f(\\{g_i\\}_{i<j})$.

Reviewer 2: ```suggestion
For a queue of transactions $Q = \{i\}_{i \ge 0}$ the gas excess $x_j$ immediately prior to execution of transaction $j \in Q$ is a monotonic, non-decreasing function of the gas usage of all preceding transactions in the queue; i.e. $x_j~:=~f(\{g_i\}_{i<j})$.
```

Reviewer 2: Apparently this suggestion is wrong - github is weird

Any sender able to pay for this upper bound (in addition to value transfers) is guaranteed to be able to pay for the actual execution cost.
Transaction _acceptance_ under worst-case cost validity is therefore a guarantee of _settlement_.

### Queue DoS protection

Worst-case cost validity only protects against DoS at the point of execution but leaves the queue vulnerable to high-limit, low-usage transactions.
For example, a malicious user could send a transfer-only transaction (21k gas) with a limit set to multiple chunks.
Although they would have to have sufficient funds to theoretically pay for all the chunks, they would never actually be charged this amount. Pushing a sufficient number of such transactions to the queue would artificially inflate the worst-case cost of other users.

Gas limits are typically set higher than expected gas used to allow for a buffer should gas estimates be imprecise.
A lower bound of 50% of the specified limit, for example, would allow for a 100% buffer on estimates without penalising the sender, while still disincentivising falsely high limits.

> #### Upper bound on queue DoS

Reviewer 2: This section should remain in the security consideration (and at the end we should plug in our planned parameters so that we can see the actual value on the worst case DoS multiplier)

With $R$ (for rate) as already defined (gas capacity per second), we introduce $g_L$, $g_U$, and $g_C$ as gas _limit_, _used_ and _charged_ per transaction, respectively, and

$$
g_C := \max(g_U, \lambda \cdot g_L)
$$

where $\lambda$ enforces a lower bound on the charge.

The actual gas excess $x_A$ has an upper bound of the worst-case excess $x_W$, both of which can be used  to calculate respective base fees $f_A$ and $f_W$ (the variable element of gas prices) from the existing exponential function:

$$
f := M \cdot \exp\left( \frac{x}{K} \right).
$$

> Mallory is attempting to maximize the DoS ratio

Reviewer 3: who?

Author: [Alice and Bob's arch nemesis](https://en.wikipedia.org/wiki/Alice_and_Bob#:~:text=A%20malicious%20attacker.)!

$$
D := \frac{f_W}{f_A}
$$

by maximizing $\Sigma_{\forall i} (g_L - g_U)_i$ to maximize $x_W - x_A$.

> [!TIP]
> Although $D$ shadows a variable in ACP-176, that one is very different to anything here so there won't be confusion.

Recall that the proportionality $p$ of capacity per second and target gas per second ($R = pT$) results in increasing excess such that

$$
x := x + \frac{g (p-1)}{p}.
$$

> If we allow the queue to have a bandwidth of $R\lambda^{-1}$ gas/second and limit its length $\tau$ seconds or a back-pressure window of $\omega = \tau R \lambda^{-1}$ gas, we can derive an upper bound on the difference in the changes to worst-case and actual gas excess:

Reviewer 2: When we include this in the Specification section, it should read something like:
> Any block that would cause $\sum_{tx \in queue} g_L > \tau R \lambda^{-1}$ is considered invalid.

> If we allow the queue to have a bandwidth of $R\lambda^{-1}$ gas/second and limit its length $\tau$ seconds or a back-pressure window of $\omega = \tau R \lambda^{-1}$ gas, we can derive an upper bound on the difference in the changes to worst-case and actual gas excess:

Reviewer 2: We should comment explicitly that we are setting the maximum bandwidth of the queue to $\frac{R}{\lambda}$ to ensure that the queue rate limits never constrain the execution speed (while minimizing the bandwidth requirements).

Reviewer 2: From the prior discussion:

Say we define:

```math
Capacity_{bandwidth}~:=~\dfrac{R}{\lambda}
```

This ensures that:

```math
g_C \geq R
```

Because (ignoring block packing inefficiencies):

```math
\begin{align}
g_C &= \sum_{\forall i}{\max\left({g_U}_i, \lambda \cdot {g_L}_i\right)}\\
g_C &\geq \sum_{\forall i}{\lambda \cdot {g_L}_i}\\
g_C &\geq \lambda \cdot \sum_{\forall i}{{g_L}_i}\\
g_C &\geq \lambda \cdot Capacity_{bandwidth}\\
g_C &\geq R
\end{align}
```

$$
\begin{align}
\Delta x_A &\ge \lambda \cdot \omega \cdot \frac{p-1}{p} \\
\Delta x_W &= \omega \cdot \frac{p-1}{p} \\
\Delta x_W - \Delta x_A &\le (1-\lambda) \cdot \omega \cdot \frac{p-1}{p} \\
&= (1-\lambda) \cdot \frac{\tau R}{\lambda} \cdot \frac{p-1}{p} \\
&= \frac{1-\lambda}{\lambda} \cdot \tau \cdot p T \cdot \frac{p-1}{p} \\
&= \frac{1-\lambda}{\lambda} \cdot \tau \cdot T \cdot (p-1).
\end{align}
$$

Note that we can express Mallory's DoS quotient as:

$$
\begin{align}
D &= \frac{f_W}{f_A} \\
&= \frac{ M \cdot \exp \left( \frac{x_W}{K} \right)}{ M \cdot \exp \left( \frac{x_A}{K} \right)} \\
& = \exp \left( \frac{x_W - x_A}{K} \right).
\end{align}
$$

When the queue is empty (i.e. the execution stream has caught up with accepted transactions), the worst-case fee estimate $f_W$ is known to be the actual base fee $f_A$; i.e. $Q = \emptyset \implies D=1$. The previous bound on $\Delta x_W - \Delta x_A$ also bounds Mallory's ability such that:

$$

> D \le \exp \left( \frac{\frac{1-\lambda}{\lambda} \cdot \tau \cdot T \cdot (p-1)}{K} \right).

Reviewer 2: In ACP-176, it is defined that: $K = 87 * T$.

We should simplify: $\frac{T}{K} = \frac{T}{87 \cdot T} = \frac{1}{87}$

| ACP | 194 |
| :--- | :--- |
| **Title** | Streaming Asynchronous Execution |
| **Author(s)** | Arran Schlosberg ([@ARR4N](https://github.com/ARR4N)), Stephen Buttolph ([@StephenButtolph](https://github.com/StephenButtolph)) |
| **Status** | Proposed, Implementable, Activated, Stale ([Discussion](POPULATED BY MAINTAINER, DO NOT SET)) |
| **Track** | Standards |

> **Note on the preamble:** The `Discussion` link is only needed for ACPs in `Proposed` and `Implementable` status.

Reviewer 3: ```suggestion
```

## Abstract

Streaming Asynchronous Execution (SAE) decouples consensus and execution by introducing a (theoretically unbounded) transaction queue upon which consensus is performed.
A concurrent execution stream is responsible for clearing the queue and reporting a delayed state root for recording by later rounds of consensus.

> Validation of transactions to be pushed to the queue is light-weight but guarantees eventual execution.

Reviewer 3: ```suggestion
Validation of transactions to be pushed to the queue is lightweight but guarantees eventual execution.
```

Without the need to compute state _roots_, such clients can eschew expensive Merkle data structures.
End users see expedited but identical transaction results.
3. Irregular stop-the-world events like database compaction are amortised over multiple blocks.
4. EOA-to-EOA transfers of value can be considered settled immediately upon being pushed to the queue by consensus.
5. Introduces bursty throughput under temporary high load, preemptively accepting transactions that may take marginally longer to execute, without a reduction in security guarantees.

### User stories

1. A sophisticated DeFi trader runs a highly optimised execution client, locally clearing the transaction queue well in advance of the network—setting the stage for HFT DeFi.
2. A custodial platform filters the queue for only those transactions sent to one of their EOAs, immediately crediting user balances.

## Specification

### Transaction queue

#### Pushing transactions

Standard, synchronous execution couples a block to the ordered set of transactions to be executed in said block.

> Let there instead be a FIFO queue of all known, valid transactions—the specifics of "validity" will be addressed later.

Reviewer 3: ```suggestion
Instead, let there be a FIFO queue of all known, valid transactions—the specifics of "validity" will be addressed later.
```
not required, but more clear imo
